cmd/nerdctl/update.go

```diff
@@ -0,0 +1,256 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/nerdctl/pkg/formatter"
+	"github.com/containerd/nerdctl/pkg/idutil/containerwalker"
+	"github.com/containerd/typeurl"
+	"github.com/docker/go-units"
+	runtimespec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+type updateOptions struct {
+	CpuPeriod uint64
+	// CPU CFS (Completely Fair Scheduler) quota. Default: 0 (not specified).
+	CpuQuota int64
+	// CPU shares (relative weight vs. other containers). Default: 0 (not specified).
+	CpuShares uint64
+	// Memory limit in bytes. Default: 0 (not specified).
+	MemoryLimitInBytes int64
+	// CpusetCpus constrains the allowed set of logical CPUs. Default: "" (not specified).
+	CpusetCpus string
+	// CpusetMems constrains the allowed set of memory nodes. Default: "" (not specified).
+	CpusetMems string
+}
+
+func newUpdateCommand() *cobra.Command {
+	var updateCommand = &cobra.Command{
+		Use:               "update [flags] CONTAINER [CONTAINER, ...]",
+		Args:              cobra.MinimumNArgs(1),
+		Short:             "Update one or more running containers",
+		RunE:              updateAction,
+		ValidArgsFunction: updateShellComplete,
+		SilenceUsage:      true,
+		SilenceErrors:     true,
+	}
+	updateCommand.Flags().SetInterspersed(false)
+	setUpdateFlags(updateCommand)
+	return updateCommand
+}
+
+func setUpdateFlags(cmd *cobra.Command) {
+	cmd.Flags().Float64("cpus", 0.0, "Number of CPUs")
+	cmd.Flags().Uint64("cpu-period", 0, "Limit CPU CFS (Completely Fair Scheduler) period")
+	cmd.Flags().Int64("cpu-quota", -1, "Limit CPU CFS (Completely Fair Scheduler) quota")
+	cmd.Flags().Uint64("cpu-shares", 0, "CPU shares (relative weight)")
+	cmd.Flags().StringP("memory", "m", "", "Memory limit")
+	cmd.Flags().String("cpuset-cpus", "", "CPUs in which to allow execution (0-3, 0,1)")
+	cmd.Flags().String("cpuset-mems", "", "MEMs in which to allow execution (0-3, 0,1)")
+}
+
+func updateAction(cmd *cobra.Command, args []string) error {
+	client, ctx, cancel, err := newClient(cmd)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+	options, err := getUpdateOption(cmd)
+	if err != nil {
+		return err
+	}
+	walker := &containerwalker.ContainerWalker{
+		Client: client,
+		OnFound: func(ctx context.Context, found containerwalker.Found) error {
+			err = updateContainer(ctx, client, found.Container.ID(), options, cmd)
+			return err
+		},
+	}
+	for _, req := range args {
+		n, err := walker.Walk(ctx, req)
+		if err != nil {
+			return err
+		} else if n == 0 {
+			return fmt.Errorf("no such container %s", req)
+		}
```

Comment:
Also error out when `n > 1`

Reply:
update can update one or more containers

Reply:
Ambiguous ID should result in an error

Reply:
ok.thank you